frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+
+const POINTS_TO_BALANCE_INIT_RATIO: u32 = 1;
+
+/// Calculate the number of points to issue from a pool as `(current_points / current_balance) *
+/// new_funds` except for some zero edge cases; see logic and tests for details.
+fn points_to_issue<T: Config>(
+	current_balance: BalanceOf<T>,
+	current_points: BalanceOf<T>,
+	new_funds: BalanceOf<T>,
+) -> BalanceOf<T> {
+	match (current_balance.is_zero(), current_points.is_zero()) {
+		(true, true) | (false, true) =>
+			new_funds.saturating_mul(POINTS_TO_BALANCE_INIT_RATIO.into()),
+		(true, false) => {
+			// The pool was totally slashed.
+
+			// This is the equivalent of `(current_points / 1) * new_funds`.
+			new_funds.saturating_mul(current_points)
+		},
+		(false, false) => {
+			// Equivalent to (current_points / current_balance) * new_funds
+			current_points
+				.saturating_mul(new_funds)
+				// We check for zero above
+				.div(current_balance)
+		},
+	}
+}
+
+// Calculate the balance of a pool to unbond as `(current_balance / current_points) *
+// delegator_points`. Returns zero if any of the inputs are zero.
+fn balance_to_unbond<T: Config>(
+	current_balance: BalanceOf<T>,
+	current_points: BalanceOf<T>,
+	delegator_points: BalanceOf<T>,
+) -> BalanceOf<T> {
+	if current_balance.is_zero() || current_points.is_zero() || delegator_points.is_zero() {
+		// There is nothing to unbond
+		return Zero::zero()
+	}
+
+	// Equivalent of (current_balance / current_points) * delegator_points
+	current_balance
+		.saturating_mul(delegator_points)
+		// We check for zero above
+		.div(current_points)
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+pub struct Delegator<T: Config> {
+	pool: T::AccountId,
+	/// The quantity of points this delegator has in the bonded pool or in a sub pool if
+	/// `Self::unbonding_era` is some.
+	points: BalanceOf<T>,
+	/// The reward pools total earnings _ever_ the last time this delegator claimed a payout.
+	/// Assuming no massive burning events, we expect this value to always be below total issuance.
+	/// This value lines up with the `RewardPool::total_earnings` after a delegator claims a
+	/// payout.
+	reward_pool_total_earnings: BalanceOf<T>,
+	/// The era this delegator started unbonding at.
+	unbonding_era: Option<EraIndex>,
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+pub struct BondedPool<T: Config> {
+	points: BalanceOf<T>,
+}
+
+impl<T: Config> BondedPool<T> {
+	/// Get the amount of points to issue for some new funds that will be bonded in the pool.
+	fn points_to_issue(&self, stash: &T::AccountId, new_funds: BalanceOf<T>) -> BalanceOf<T> {
+		let bonded_balance = T::StakingInterface::bonded_balance(stash).unwrap_or(Zero::zero());
```

Comment:
The current design is that we identify a pool by its stash/controller account. So the map key for the sub pools, reward pool, and bonded pool is the stash for the pool. This way the delegator just needs to store the stash of the pool it belongs to. 

This design also makes slashing more efficient since we can just key into any of the pool maps with the stash to slash to see if its a pool. If we used a simple u32 id we would need to maintain a map of pool_stash_account_id => to PoolId in order to check if a nominator is a pool

One thing I was having the pool keep its account id, but annotate that field with codec skip. Then we can write a custom get for the pool that takes the pool's account_id, fetches the pool, and then inserts the account_id into into the pool struct. It would look very similar to: 

https://github.com/paritytech/substrate/blob/7c6342047c992b6f3fa917d0d0448eb7e89afa6c/frame/bags-list/src/list/mod.rs#L570-L575

Reply:
See 4a67640166a4092717ee8f34a26794d1187d2bf6 & 1fc884b3d5680c07f3c628f03cae08bec8cd06ba - BondedPool now contains its account id, but only the bonded pool points are in storage.

Reply:
> If we used a simple u32 id we would need to maintain a map of pool_stash_account_id => to PoolId in order to check if a nominator is a pool.

This is a good point and I agree with you now.